Add "Full Formatting" Option

t-dub opened the pull request:
# Enhancement Adding "Full Format" Option

`README` delta should clearly explain the new feature.

This addresses a desire I had which was to be able to copy code formatted as markdown in addition to a reference to the "remote" version.  My use case is for sharing code with context via Slack to coworkers.  In order to achieve that, I needed to be able to have a bit more control over the output.

This introduces a new `output_formats` format which takes an optional `{code}` token.  That allows for both the case I described above but also for _omitting_ the code altogether to quickly be able to pull a reference to the remote URL line.  These are definitely secondary to the original intent here, but once I started using this plugin, I realized this was within reach.

This is intended to be a fully backward-compatible, non-breaking change.  I'm happy to rework it in such a way that is "cleaner" but breaking if desired, but for an unsolicited PR, this seemed the best way to start.

t-dub committed:
feat: allow format control of output

- Introduce {code} output token
- Introduce output_formats alternative to formats as mechanism for
  specifying complete output format

t-dub committed:
test: update tests to account for code token

t-dub committed:
docs: update README to include code

t-dub committed:
fix: let empty table merge

t-dub committed:
format

zhisme commented on lua/copy_with_context/config.lua:
```diff
@@ -8,9 +8,13 @@ M.options = {
     relative = "<leader>cy",
     absolute = "<leader>cY",
   },
+  -- Formats: code is automatically prepended with a newline
```
I don't like the naming `code`. Seems ambiguos to me, not only code can be copied. When at first I've read `code` I thought it is something like reserved word and expected to compare with `markdown` or other formats. 

So I would go with something straightforward like `copied_text` instead of `code`

t-dub-lz replied:
Yeah that makes a lot more sense

t-dub replied:
Yeah, that makes more sense

zhisme reviewed (changes requested):
Thank you @t-dub for the contribution! I think this is useful enhancement, plugin users will find it useful for sure (I do already)

I would like to merge this PR after requested changes, also please check test-coverage reported

t-dub committed:
feat: switch from code to copied_text name

t-dub committed:
docs: add example for precedence:

t-dub committed:
test: added test for more config validation cases

t-dub committed:
lint & format

t-dub committed:
feat: add mapping descriptions

Added automatically-derived desc to the keymaps

t-dub commented:
@zhisme OK, I think I got everything you raised if you wouldn't mind giving it another look.

I also snuck in another change, which is to add automatically-derived descriptions to the keymaps so that things like which-key.nvim will have something to display alongside the mapping.

zhisme commented on README.md:
```diff
@@ -218,6 +218,7 @@ You can use the following variables in format strings:
 - `{line}` - Line number or range (e.g., "42" or "10-20")
 - `{linenumber}` - Alias for `{line}`
 - `{remote_url}` - Repository URL (GitHub, GitLab, Bitbucket)
+- `{copied_text}` - The selected text (used with `output_formats`)READ
```
some extra text `)READ`

zhisme commented on README.md:
````diff
@@ -243,6 +244,45 @@ require('copy_with_context').setup({
+    default = '# {filepath}:{line}',  -- Code is auto-prepended
+  },
+  output_formats = {
+    markdown = "```{code}```\n\n{remote_url}",  -- Code token must be specified
````
I think this is enough. quite comprehensive

zhisme reviewed (changes requested):
Agreed on `which-key.nvim`, that would be cool!

Regarding the code one small bug I spotted, once solved we are ready to merge

zhisme committed:
fix: validate {copied_text} usage in formats vs output_formats

- output_formats must contain {copied_text} to include the copied code
- formats cannot contain {copied_text} (auto-prepended, would cause
duplication)
